### PR TITLE
Add 'hibernate.discriminator.ignore_explicit_for_joined' property

### PR DESCRIPTION
**Pull request by leonardowestphal**
Add 'hibernate.discriminator.ignore_explicit_for_joined' property.

### CODE REVIEW

**Comment by leonardowestphal** [comment]
This property is important for legacy projects. I'm finishing a migration of an application that needs it.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmConfigPersistenceUnit.java [hunk 1]
```diff
@@ -443,4 +451,24 @@ public boolean isAnyPropertySet() {
...
+     */
+    @ConfigGroup
+    public static class HibernateOrmConfigPersistenceUnitDiscriminator {
+        /**
+         * Existing applications rely (implicitly or explicitly) on Hibernate ignoring any DiscriminatorColumn declarations on
+         * joined inheritance hierarchies. This setting allows these applications to maintain the legacy behavior of
+         * DiscriminatorColumn annotations being ignored when paired with joined inheritance.
+         */
+        @ConfigItem(defaultValue = "false")
+        public Optional<Boolean> ignoreExplicitForJoined;
```
Comment (gastaldi) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmConfigPersistenceUnit.java:468]: You don't need to use `Optional` when you have a default value

File: extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/ignore_explicit_for_joined/IgnoreExplicitForJoinedDefaultValueWithMultiplePUsTest.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    @Inject
+    EntityManager emUsers;
+
+    @PersistenceUnit("inventory")
+    @Inject
+    EntityManager emInventory;
+
+    @BeforeEach
+    public void activateRequestContext() {
+        Arc.container().requestContext().activate();
```
Comment (gastaldi) [inline extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/ignore_explicit_for_joined/IgnoreExplicitForJoinedDefaultValueWithMultiplePUsTest.java:42]: I think you can simplify this by using `@ActivateRequestContext` in your `@Test`? (@mkouba might confirm if that works)

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java [hunk 3]
```diff
@@ -994,6 +994,12 @@ private static void producePersistenceUnitDescriptorFromConfig(
             storageEngineCollector.add(persistenceUnitConfig.dialect.storageEngine.get());
         }
 
+        // Discriminator Column
+        persistenceUnitConfig.discriminator.ignoreExplicitForJoined.ifPresent(
```
Comment (gastaldi) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java:998]: Avoid using lambdas here (for performance purposes)

File: extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/ignore_explicit_for_joined/IgnoreExplicitForJoinedDefaultValueWithMultiplePUsTest.java (hunk 2 above)
Comment (mkouba) [reply extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/ignore_explicit_for_joined/IgnoreExplicitForJoinedDefaultValueWithMultiplePUsTest.java:42]: Yep, the `@ActivateRequestContext` should work.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java (hunk 3 above)
Comment (gsmet) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java:998]: This is in build time code, we allow using lambda there. We enforce this rule only in runtime code.

**Comment by gsmet** [comment]
@Sanne what's your take on that? Our initial approach was to limit the number of properties we expose to avoid having to support the whole range of Hibernate ORM combinations.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmConfigPersistenceUnit.java (hunk 1 above)
Comment (leonardowestphal) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmConfigPersistenceUnit.java]: Done!

File: extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/ignore_explicit_for_joined/IgnoreExplicitForJoinedDefaultValueWithMultiplePUsTest.java (hunk 2 above)
Comment (leonardowestphal) [reply extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/ignore_explicit_for_joined/IgnoreExplicitForJoinedDefaultValueWithMultiplePUsTest.java]: Done!

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java (hunk 3 above)
Comment (leonardowestphal) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java]: Done!

**Comment by leonardowestphal** [comment]
> @Sanne what's your take on that? Our initial approach was to limit the number of properties we expose to avoid having to support the whole range of Hibernate ORM combinations.

I understand that you don't want to support all the properties, but I need this one specifically and also this one https://github.com/quarkusio/quarkus/pull/17509 to finally finish a migration from a Jboss EAP legacy project. Without it I will be wasting hundreds of hours of work! As it's a property supported by Hibernate I think it shouldn't be a problem to add it as it won't affect the behavior of Quarkus and the hibernate-orm extension.

**Comment by Sanne** [comment]
hi @leonardowestphal sorry I'm slowly catching up with all your issues :) - spent some days deeply diving in a complex issue.

I'll share a little secret with you: if you set add an `hibernate.properties` resource to your project, you can add properties in there like

```
hibernate.discriminator.ignore_explicit_for_joined=true
```

and these will be applied on the Hibernate ORM instances (all of them) that you're initializing in Quarkus.

N.B. this isn't documented because we will actually remove it soon. In 2.0 (main branch) it was already removed, but if it helps you moving forward with a migration you can take advantage of it - just please remember we need to keep working on issues like this one so that you won't be stuck and unable to migrate to 2.0.

**Review by Sanne — APPROVED** [review]
Great patch! I love PRs with such thourough testing.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java [hunk 4]
```diff
@@ -3,10 +3,7 @@
 import static io.quarkus.deployment.annotations.ExecutionTime.RUNTIME_INIT;
 import static io.quarkus.deployment.annotations.ExecutionTime.STATIC_INIT;
 import static io.quarkus.hibernate.orm.deployment.HibernateConfigUtil.firstPresent;
-import static org.hibernate.cfg.AvailableSettings.JPA_SHARED_CACHE_MODE;
-import static org.hibernate.cfg.AvailableSettings.USE_DIRECT_REFERENCE_CACHE_ENTRIES;
-import static org.hibernate.cfg.AvailableSettings.USE_QUERY_CACHE;
-import static org.hibernate.cfg.AvailableSettings.USE_SECOND_LEVEL_CACHE;
+import static org.hibernate.cfg.AvailableSettings.*;
```
Comment (machi1990) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java:6]: Let's avoid wildcard imports.

Comment (Sanne) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java:6]: good point

**Review by Sanne — CHANGES_REQUESTED** [review]
Changing to "request changes" to address @machi1990 point.

+1 to expose this property.

**Comment by leonardowestphal** [comment]
> hi @leonardowestphal sorry I'm slowly catching up with all your issues :) - spent some days deeply diving in a complex issue.
> 
> I'll share a little secret with you: if you add an `hibernate.properties` resource to your project, you can add properties in there like
> 
> ```
> hibernate.discriminator.ignore_explicit_for_joined=true
> ```
> 
> and these will be applied on the Hibernate ORM instances (all of them) that you're initializing in Quarkus.
> 
> N.B. this isn't documented because we will actually remove it soon. In 2.0 (main branch) it was already removed, but if it helps you moving forward with a migration you can take advantage of it - just please remember we need to keep working on issues like this one so that you won't be stuck and unable to migrate to 2.0.

Thanks for sharing this tip. This will help us continue the project!

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java (hunk 4 above)
Comment (leonardowestphal) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java]: Done!

Comment (machi1990) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java]: Thanks.

**Review by machi1990 — APPROVED** [review]

**Review by Sanne — APPROVED** [review]

**Comment by leonardowestphal** [comment]
In which version do you intend to merge this PR?

**Comment by Sanne** [comment]
@leonardowestphal we always merge PRs in `main`.  After that, things can be suggested for backporting - but we typically will do so later, to give things time to be battle tested.

At this point "main" means "after 2.0" as 2.0 was already branched. So if you need this in some 1.x version, we'd need to backport it two steps. I suppose you need this in some 1.x?

**Comment by leonardowestphal** [comment]
> @leonardowestphal we always merge PRs in `main`. After that, things can be suggested for backporting - but we typically will do so later, to give things time to be battle tested.
> 
> At this point "main" means "after 2.0" as 2.0 was already branched. So if you need this in some 1.x version, we'd need to backport it two steps. I suppose you need this in some 1.x?

Yes I need it in version 1.13.x.

**Pull request merged**